# Part 09. 현재 작업 폴더와 필수 CSV 확인

## 38. 프로젝트 루트 확인

In [130]:
from pathlib import Path

project_root = Path.cwd()

if project_root.name == "notebooks":
    project_root = project_root.parent

data_dir = project_root / "data" / "raw"

print("프로젝트 루트:", project_root)
print("데이터 폴더:", data_dir)
print("데이터 폴더 존재:", data_dir.exists())

프로젝트 루트: c:\dev\ai-data-analysis
데이터 폴더: c:\dev\ai-data-analysis\data\raw
데이터 폴더 존재: True


## 38. 필수 파일 확인

In [131]:

required_files = [

    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",

]

for file_name in required_files:
    file_path = data_dir / file_name

    print(file_name, file_path.exists(), file_path.stat().st_size if file_path.exists() else 0)

customers.csv True 5641
products.csv True 4103
orders.csv True 11693
order_items.csv True 15335


## 40. 데이터 불러오기

In [132]:
import pandas as pd

customers = pd.read_csv(data_dir / "customers.csv")
products = pd.read_csv(data_dir / "products.csv")
orders = pd.read_csv(data_dir / "orders.csv")      # 영수증과 유사함
order_items = pd.read_csv(data_dir / "order_items.csv")

## 41. DataFrame 사전 만들기

In [133]:
datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

print(datasets)

{'customers':      customer_id name gender  age city signup_date
0              1  김수민      F   19   광주  2024-07-15
1              2  김정호      F   32   대구  2025-11-28
2              3  이경수      F   61   성남  2024-07-08
3              4  조영호      F   55   울산  2026-05-09
4              5  이예원      F   19   부산  2024-10-09
..           ...  ...    ...  ...  ...         ...
145          146  김숙자      M   61   성남  2026-01-18
146          147  이정남      M   19   부산  2025-03-09
147          148  오도현      M   29   고양  2026-07-11
148          149  김정자      M   20   부산  2024-11-14
149          150  조미영      M   40   대전  2025-12-30

[150 rows x 6 columns], 'products':     product_id product_name category   price
0            1  전자기기 상품 001     전자기기  160000
1            2    도서 상품 002       도서   34000
2            3  전자기기 상품 003     전자기기  152000
3            4  생활용품 상품 004     생활용품   70000
4            5    식품 상품 005       식품  186000
..         ...          ...      ...     ...
95          96  생활용품 상

## 42. 크기 요약

In [134]:
shape_summary = pd.DataFrame(
    [
        {
            "dataset": name,
            "rows": df.shape[0],
            "columns": df.shape[1],
        }
        for name, df in datasets.items()
    ]
)

print(shape_summary)

       dataset  rows  columns
0    customers   150        6
1     products   100        4
2       orders   301        5
3  order_items   764        5


## 43. 한 행의 의미 기록

## 데이터셋별 한 행의 의미

- customers: 고객 한 명

- products: 상품 한 개

- orders: 주문 한 건

- order_items: 주문에 포함된 상품 한 항목

## 44. 처음과 마지막 행 확인

In [135]:
for name, df in datasets.items():

    print(f"\n[{name}] 앞부분")
    display(df.head())
    print(f"[{name}] 뒷부분")
    display(df.tail())


[customers] 앞부분


,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-07-15
1,2,김정호,F,32,대구,2025-11-28
2,3,이경수,F,61,성남,2024-07-08
3,4,조영호,F,55,울산,2026-05-09
4,5,이예원,F,19,부산,2024-10-09


[customers] 뒷부분


,customer_id,name,gender,age,city,signup_date
145,146,김숙자,M,61,성남,2026-01-18
146,147,이정남,M,19,부산,2025-03-09
147,148,오도현,M,29,고양,2026-07-11
148,149,김정자,M,20,부산,2024-11-14
149,150,조미영,M,40,대전,2025-12-30



[products] 앞부분


,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000


[products] 뒷부분


,product_id,product_name,category,price
95,96,생활용품 상품 096,생활용품,112000
96,97,전자기기 상품 097,전자기기,154000
97,98,스포츠 상품 098,스포츠,10000
98,99,뷰티 상품 099,뷰티,200000
99,100,도서 상품 100,도서,102000



[orders] 앞부분


,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-06-02,card,completed
1,2,77,2025-08-18,naver_pay,cancelled
2,3,138,2025-12-15,bank_transfer,cancelled
3,4,57,2026-02-25,kakao_pay,cancelled
4,5,125,2026-01-16,card,cancelled


[orders] 뒷부분


,order_id,customer_id,order_date,payment_method,order_status
296,297,22,2026-02-10,bank_transfer,completed
297,298,64,2026-01-20,naver_pay,cancelled
298,299,135,2025-09-03,bank_transfer,completed
299,300,96,2026-03-23,kakao_pay,cancelled
300,301,300,2026-03-23,kakao_pay,cancelled



[order_items] 앞부분


,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


[order_items] 뒷부분


,order_item_id,order_id,product_id,quantity,unit_price
759,760,297,42,4,28000
760,761,298,40,4,174000
761,762,299,8,2,189000
762,763,299,12,4,175000
763,764,300,59,5,32000


## 45. 컬럼명 확인

In [136]:
for name, df in datasets.items():
    print(name, df.columns.tolist())

customers ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
products ['product_id', 'product_name', 'category', 'price']
orders ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
order_items ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']


## 46. 컬럼명 공백 점검

In [137]:
for name, df in datasets.items():

    suspicious = [
        col for col in df.columns
        if col != col.strip()
    ]

    print(name, suspicious)

customers []
products []
orders []
order_items []


## 47. info() 실행

In [138]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    df.info()


===== customers =====
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB

===== products =====
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   product_id    100 non-null    int64
 1   product_name  100 non-null    str  
 2   category      100 non-null    str  
 3   price         100 non-null    int64
dtypes: int64(2), str(2)
memory usage: 3.3 KB

===== orders =====
<class 'pandas.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 5 columns):

## 48. dtype 요약표

In [139]:
dtype_rows = []
for name, df in datasets.items():
    for column, dtype in df.dtypes.items():
        dtype_rows.append(
            {
                "dataset": name,
                "column": column,
                "dtype": str(dtype),
            }
        )
dtype_summary = pd.DataFrame(dtype_rows)
dtype_summary

,dataset,column,dtype
0,customers,customer_id,int64
1,customers,name,str
2,customers,gender,str
3,customers,age,int64
4,customers,city,str
5,customers,signup_date,str
6,products,product_id,int64
7,products,product_name,str
8,products,category,str
9,products,price,int64


## 49. 날짜형 후보 확인

In [140]:
date_columns = {
    "customers": ["signup_date"],
    "orders": ["order_date"],
}

for dataset_name, columns in date_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        converted = pd.to_datetime(
                   # pd = pandas
            df[column],
            errors="coerce"
        )
        print(
            dataset_name,
            column,
            "원본 dtype:", df[column].dtype,
            "변환 실패:", converted.isna().sum()
        )

customers signup_date 원본 dtype: str 변환 실패: 0
orders order_date 원본 dtype: str 변환 실패: 0


## 50. 결측치 개수

In [141]:
for name, df in datasets.items():
    print(f"\n[{name}]")
    display(df.isna().sum().to_frame("missing_count"))


[customers]


,missing_count
customer_id,0
name,0
gender,0
age,0
city,0
signup_date,0



[products]


,missing_count
product_id,0
product_name,0
category,0
price,0



[orders]


,missing_count
order_id,0
customer_id,0
order_date,0
payment_method,0
order_status,0



[order_items]


,missing_count
order_item_id,0
order_id,0
product_id,0
quantity,0
unit_price,0


## 51. 결측치 비율

결측치 비율 중요,, 결측치 높으면 결측 살릴 수 있을지 통계/분석 문제 없을지 고민

In [142]:
missing_summaries = []
for name, df in datasets.items():
    summary = pd.DataFrame(
        {
            "missing_count": df.isna().sum(),
            "missing_rate_pct": (
                df.isna().mean() * 100
            ).round(2),
        }
    )
    summary.insert(0, "dataset", name)
    summary.insert(1, "column", summary.index)
    missing_summaries.append(summary.reset_index(drop=True))
    
missing_summary = pd.concat(
    missing_summaries,
    ignore_index=True
)

print(missing_summary)

        dataset          column  missing_count  missing_rate_pct
0     customers     customer_id              0               0.0
1     customers            name              0               0.0
2     customers          gender              0               0.0
3     customers             age              0               0.0
4     customers            city              0               0.0
5     customers     signup_date              0               0.0
6      products      product_id              0               0.0
7      products    product_name              0               0.0
8      products        category              0               0.0
9      products           price              0               0.0
10       orders        order_id              0               0.0
11       orders     customer_id              0               0.0
12       orders      order_date              0               0.0
13       orders  payment_method              0               0.0
14       orders    order_

## 52. 전체 행 중복

In [143]:
for name, df in datasets.items():
    print(name, "전체 중복:", df.duplicated().sum())
                      # 동명이인이 같은 시간에 들어가는 등 중복이 생길 수도 있다.
                      # 그래서 duplicated 사용

customers 전체 중복: 0
products 전체 중복: 0
orders 전체 중복: 0
order_items 전체 중복: 0


## 53. 주요 ID 점검

In [144]:
primary_keys = {
    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",
}

id_check_rows = []

for name, key in primary_keys.items():
    df = datasets[name]
    id_check_rows.append(
        {
            "dataset": name,
            "key": key,
            "missing": df[key].isna().sum(),
            "duplicates": df[key].duplicated().sum(),
            "unique": df[key].nunique(),
            "rows": len(df),
        }
    )
id_check_summary = pd.DataFrame(id_check_rows)

id_check_summary

,dataset,key,missing,duplicates,unique,rows
0,customers,customer_id,0,0,150,150
1,products,product_id,0,0,100,100
2,orders,order_id,0,0,301,301
3,order_items,order_item_id,0,0,764,764


## 54. 발견 내용을 기록합니다

## 데이터 품질 발견 기록

| 데이터셋 | 컬럼 | 문제 종류 | 건수 | 해석 | 처리 여부 |

|---|---|---|---:|---|---|

| | | | | | Chapter 05에서 결정 |

| 데이터셋        | 컬럼            | 문제 종류     | 건수 | 해석                   | 처리 여부  |
| ----------- | ------------- | --------- | -: | -------------------- | ------ |
| customers   | 전체 컬럼         | 결측치       | 0건 | 모든 컬럼에 결측치가 없음       | 처리 불필요 |
| products    | 전체 컬럼         | 결측치       | 0건 | 모든 컬럼에 결측치가 없음       | 처리 불필요 |
| orders      | 전체 컬럼         | 결측치       | 0건 | 모든 컬럼에 결측치가 없음       | 처리 불필요 |
| order_items | 전체 컬럼         | 결측치       | 0건 | 모든 컬럼에 결측치가 없음       | 처리 불필요 |
| customers   | 전체 행          | 중복 데이터    | 0건 | 중복된 행이 존재하지 않음       | 처리 불필요 |
| products    | 전체 행          | 중복 데이터    | 0건 | 중복된 행이 존재하지 않음       | 처리 불필요 |
| orders      | 전체 행          | 중복 데이터    | 0건 | 중복된 행이 존재하지 않음       | 처리 불필요 |
| order_items | 전체 행          | 중복 데이터    | 0건 | 중복된 행이 존재하지 않음       | 처리 불필요 |
| customers   | customer_id   | 기본키 결측/중복 | 0건 | 기본키가 모두 고유하며 결측치가 없음 | 처리 불필요 |
| products    | product_id    | 기본키 결측/중복 | 0건 | 기본키가 모두 고유하며 결측치가 없음 | 처리 불필요 |
| orders      | order_id      | 기본키 결측/중복 | 0건 | 기본키가 모두 고유하며 결측치가 없음 | 처리 불필요 |
| order_items | order_item_id | 기본키 결측/중복 | 0건 | 기본키가 모두 고유하며 결측치가 없음 | 처리 불필요 |


# Part 14. 기본 통계·범주형 빈도·날짜 범위 확인

 

## 55. 숫자형 기본 통계

In [145]:
print(customers[["age"]].describe())
print(products[["price"]].describe())
print(order_items[["quantity", "unit_price"]].describe())

              age
count  150.000000
mean    42.086667
std     15.613166
min     19.000000
25%     29.000000
50%     40.000000
75%     57.000000
max     69.000000
               price
count     100.000000
mean   110040.000000
std     56433.910574
min      5000.000000
25%     65750.000000
50%    112000.000000
75%    161000.000000
max    200000.000000
         quantity     unit_price
count  764.000000     764.000000
mean     3.053665  108561.518325
std      1.410873   56996.770604
min      1.000000    5000.000000
25%      2.000000   62000.000000
50%      3.000000  111000.000000
75%      4.000000  161250.000000
max      5.000000  200000.000000


# 56. 범주형 값 확인

In [146]:
category_columns = {
    "customers": ["gender", "city"],
    "products": ["category"],
    "orders": ["payment_method", "order_status"],
}

for dataset_name, columns in category_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        print(f"\n[{dataset_name}.{column}]")

        display(
            df[column]
            .value_counts(dropna=False)
            .to_frame("count")
        )


[customers.gender]


,count
gender,
F,84
M,66



[customers.city]


,count
city,
성남,21
광주,17
부산,16
대구,15
서울,15
울산,14
인천,14
대전,14
수원,13



[products.category]


,count
category,
스포츠,19
전자기기,17
생활용품,16
뷰티,16
도서,14
패션,11
식품,7



[orders.payment_method]


,count
payment_method,
kakao_pay,80
naver_pay,77
bank_transfer,74
card,70



[orders.order_status]


,count
order_status,
completed,184
cancelled,65
refunded,52


## 57. 날짜 범위 확인

In [147]:
signup_dates = pd.to_datetime(
    customers["signup_date"],
    errors="coerce"
)

order_dates = pd.to_datetime(
    orders["order_date"],
    errors="coerce"
)

print("가입일 변환 실패:", signup_dates.isna().sum())
print("가입일 범위:", signup_dates.min(), "~", signup_dates.max())
print("주문일 변환 실패:", order_dates.isna().sum())
print("주문일 범위:", order_dates.min(), "~", order_dates.max())

가입일 변환 실패: 0
가입일 범위: 2023-08-09 00:00:00 ~ 2026-07-23 00:00:00
주문일 변환 실패: 0
주문일 범위: 2025-08-04 00:00:00 ~ 2026-08-03 00:00:00


## 58. 날짜 관계 점검 준비

주문일과 고객 가입일을 직접 비교하려면 두 데이터를 연결해야 합니다.

이번 장에서는 구조와 키가 연결 가능한지 확인하고, 실제 병합과 분석은 이후 장에서 본격적으로 진행합니다.

# Part 15. 파일 간 연결 무결성·반복 점검 함수·LLM 구조 요약

## 59. 참조하지 못한 고객 ID

In [148]:
unknown_customer_ids = sorted(
    set(orders["customer_id"].dropna())
    - set(customers["customer_id"].dropna())
)

print("존재하지 않는 customer_id:", unknown_customer_ids)

# oders 파일에 300고객 새로 만들어서 참조하지 못한 고객,, 

존재하지 않는 customer_id: [300]


## 60. 참조하지 못한 주문 ID

In [149]:
unknown_order_ids = sorted(
    set(order_items["order_id"].dropna())
    - set(orders["order_id"].dropna())
)

print("존재하지 않는 order_id:", unknown_order_ids)

존재하지 않는 order_id: []


## 61. 참조하지 못한 상품 ID

In [150]:
unknown_product_ids = sorted(
    set(order_items["product_id"].dropna())
    - set(products["product_id"].dropna())
)

print("존재하지 않는 product_id:", unknown_product_ids)

존재하지 않는 product_id: []


## 62. 반복 점검 함수

In [151]:
def inspect_dataframe(

    name: str,
    df: pd.DataFrame,
    primary_key: str | None = None,

) -> dict:

    result = {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
        "column_names": df.columns.tolist(),
        "dtypes": {

            column: str(dtype)
            for column, dtype in df.dtypes.items()

        },

        "missing_total": int(df.isna().sum().sum()),
        "duplicate_rows": int(df.duplicated().sum()),

    }

    if primary_key is not None:

        result["primary_key"] = primary_key
        result["key_missing"] = int(
            df[primary_key].isna().sum()

        )

        result["key_duplicates"] = int(
            df[primary_key].duplicated().sum()

        )

        result["key_unique"] = int(
            df[primary_key].nunique()

        )

    return result

In [152]:
inspection_results = []

for name, df in datasets.items():
    inspection_results.append(
        inspect_dataframe(
            name,
            df,
            primary_keys[name],
        )
    )

print(inspection_results)

[{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status': 'str'}, 'mi

## 63. LLM에 전달할 안전한 구조 요약

In [153]:
safe_schema_summary = {

    "datasets": inspection_results,
    "relationships": {
        "orders.customer_id": {
            "references": "customers.customer_id",
            "unknown_count": len(unknown_customer_ids),
        },

        "order_items.order_id": {
            "references": "orders.order_id",
            "unknown_count": len(unknown_order_ids),
        },

        "order_items.product_id": {
            "references": "products.product_id",
            "unknown_count": len(unknown_product_ids),
        },
    },
}

print(safe_schema_summary)

{'datasets': [{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status'

## 64. LLM 구조 점검 프롬프트

 

나는 온라인 쇼핑몰 데이터 분석을 준비하고 있습니다.

아래 정보는 원본 행이 아니라 구조 점검 요약입니다.

{'datasets': [{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'order_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 301}, {'dataset': 'order_items', 'rows': 764, 'columns': 5, 'column_names': ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price'], 'dtypes': {'order_item_id': 'int64', 'order_id': 'int64', 'product_id': 'int64', 'quantity': 'int64', 'unit_price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'order_item_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 764}], 'relationships': {'orders.customer_id': {'references': 'customers.customer_id', 'unknown_count': 1}, 'order_items.order_id': {'references': 'orders.order_id', 'unknown_count': 0}, 'order_items.product_id': {'references': 'products.product_id', 'unknown_count': 0}}}

다음을 검토해 주세요.

1. 각 데이터셋에서 한 행이 무엇을 의미할 가능성이 높은지 설명해 주세요.

| 데이터셋            | 한 행이 의미하는 것                                                             |
| --------------- | ----------------------------------------------------------------------- |
| **customers**   | **고객 1명의 정보**를 의미합니다. (고객 ID, 이름, 성별, 나이, 거주 도시, 가입일)                   |
| **products**    | **상품 1개의 정보**를 의미합니다. (상품 ID, 상품명, 카테고리, 가격)                            |
| **orders**      | **주문 1건의 정보**를 의미합니다. (주문 ID, 주문한 고객, 주문일, 결제수단, 주문 상태)                 |
| **order_items** | **주문에 포함된 상품 1개**를 의미합니다. (주문 상세 정보로, 어떤 주문에 어떤 상품이 몇 개 얼마에 판매되었는지 나타냄) |


2. 주요 ID와 외래 키를 구분해 주세요.
| 데이터셋        | 컬럼              | 구분          | 설명                                        |
| ----------- | --------------- | ----------- | ----------------------------------------- |
| customers   | `customer_id`   | **기본키(PK)** | 고객을 고유하게 식별하는 ID                          |
| products    | `product_id`    | **기본키(PK)** | 상품을 고유하게 식별하는 ID                          |
| orders      | `order_id`      | **기본키(PK)** | 주문을 고유하게 식별하는 ID                          |
| order_items | `order_item_id` | **기본키(PK)** | 주문 상세를 고유하게 식별하는 ID                       |
| orders      | `customer_id`   | **외래키(FK)** | `customers.customer_id`를 참조하여 주문한 고객을 나타냄 |
| order_items | `order_id`      | **외래키(FK)** | `orders.order_id`를 참조하여 어떤 주문인지 나타냄       |
| order_items | `product_id`    | **외래키(FK)** | `products.product_id`를 참조하여 어떤 상품인지 나타냄   |


customers
┌──────────────┐
│ customer_id(PK) │
└──────┬───────┘
       │
       │
orders
┌────────────┐
│ order_id(PK) │
│ customer_id(FK) │
└──────┬─────┘
       │
       │
order_items
┌─────────────────┐
│ order_item_id(PK) │
│ order_id(FK)      │
│ product_id(FK)    │
└──────┬──────────┘
       │
       │
products
┌─────────────┐
│ product_id(PK) │
└─────────────┘

3. 즉시 삭제하면 안 되는 결측치·중복 사례를 알려 주세요.

| 사례                     | 이유                                               |
| ---------------------- | ------------------------------------------------ |
| 기본키(PK)의 중복            | 데이터 입력 오류인지, 실제 중복인지 확인해야 함                      |
| 외래키(FK)의 불일치           | 관련 데이터가 누락되었거나 잘못 입력되었을 수 있으므로 원인 분석 후 처리해야 함    |
| 중요한 컬럼의 결측치            | 삭제하면 데이터 손실이 발생할 수 있으므로 대체(평균, 최빈값 등) 여부를 검토해야 함 |
| 주문 상세(order_items)의 중복 | 동일 상품을 여러 번 주문한 정상 데이터일 수도 있으므로 확인이 필요함          |


4. 추가로 점검할 구조 문제를 우선순위대로 제안해 주세요.

|  우선순위 | 점검 항목              | 이유                                                                                      |
| ----: | ------------------ | --------------------------------------------------------------------------------------- |
| **1** | **외래키 무결성 확인**     | `orders.customer_id`에서 존재하지 않는 고객 ID가 1건 발견되었으므로 가장 먼저 확인해야 합니다.                        |
| **2** | **데이터 타입 확인**      | `order_date`, `signup_date`가 문자열(`str`)이므로 날짜(`datetime`)로 변환 가능한지 확인해야 합니다.            |
| **3** | **범주형 데이터 일관성 확인** | `gender`, `city`, `payment_method`, `order_status`, `category`에 오타나 불필요한 공백이 있는지 확인합니다. |
| **4** | **수치 데이터 범위 확인**   | `age`, `price`, `quantity`, `unit_price`에 음수나 비정상적으로 큰 값이 있는지 확인합니다.                    |
| **5** | **날짜 순서 확인**       | `signup_date`보다 `order_date`가 더 이전인 데이터가 있는지 확인합니다. 가입 전 주문은 비정상 데이터일 가능성이 있습니다.        |


5. 실제 데이터에 없는 컬럼이나 관계를 가정하지 마세요.

|  우선순위 | 점검 항목                                                                    | 근거                                             |
| ----: | ------------------------------------------------------------------------ | ---------------------------------------------- |
| **1** | `orders.customer_id`의 외래키 무결성 확인                                         | `customers.customer_id`에 존재하지 않는 값이 **1건** 확인됨 |
| **2** | `order_date`, `signup_date`의 데이터 타입 확인                                   | 현재 문자열(`str`)이므로 날짜 형식으로 변환 가능한지 확인 필요         |
| **3** | `gender`, `city`, `payment_method`, `order_status`, `category` 값의 일관성 확인 | 동일한 의미의 값이 서로 다르게 입력되었는지(오타, 공백 등) 확인          |
| **4** | `age`, `price`, `quantity`, `unit_price`의 값 범위 확인                        | 음수 또는 비정상적인 값이 있는지 확인                          |


6. 각 제안에 대해 확인할 pandas 코드의 목적만 설명해 주세요.

| 점검 항목                                                                               | pandas 코드의 목적                                                            |
| ----------------------------------------------------------------------------------- | ------------------------------------------------------------------------ |
| **외래키 무결성 확인 (`orders.customer_id`)**                                               | `orders`의 `customer_id`가 `customers`에 모두 존재하는지 확인하여 참조 무결성 오류를 찾기 위함입니다. |
| **데이터 타입 확인 (`order_date`, `signup_date`)**                                         | 날짜 컬럼이 올바른 형식으로 변환 가능한지 확인하여 날짜 분석이 가능한 상태인지 점검하기 위함입니다.                 |
| **범주형 데이터 일관성 확인 (`gender`, `city`, `payment_method`, `order_status`, `category`)** | 같은 의미의 값이 서로 다른 형태(오타, 대소문자, 공백 등)로 입력되었는지 확인하기 위함입니다.                   |
| **수치 데이터 범위 확인 (`age`, `price`, `quantity`, `unit_price`)**                         | 음수나 비정상적으로 큰 값 등 데이터 입력 오류가 있는지 확인하기 위함입니다.                              |


개인정보나 원본 고객 행은 제공하지 않았습니다.

# 65. LLM 답변 검증표

| 검증 항목       | 결과   | 근거                                                                                                                                                     |
| ----------- | ---- | ------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **컬럼 존재**   | ✅ 통과 | 실제 컬럼(`customer_id`, `product_id`, `order_id`, `order_item_id`, `order_date`, `signup_date` 등)만 사용했습니다.                                                |
| **한 행의 의미** | ✅ 통과 | `customers`는 고객 1명, `products`는 상품 1개, `orders`는 주문 1건, `order_items`는 주문 상품 1개로 설명했습니다.                                                               |
| **ID 해석**   | ✅ 통과 | ID를 수치 분석 변수로 해석하지 않고 식별자(기본키/외래키)로 설명했습니다.                                                                                                            |
| **중복 해석**   | ✅ 통과 | 외래키(`customer_id`, `order_id`, `product_id`)는 여러 번 나타날 수 있으며, 이를 중복 오류로 판단하지 않았습니다.                                                                    |
| **결측 해석**   | ✅ 통과 | 결측치 삭제를 권하지 않았으며, 실제 데이터에서도 결측치가 없음을 확인했습니다.                                                                                                           |
| **관계**      | ✅ 통과 | 실제 관계인 `orders.customer_id → customers.customer_id`, `order_items.order_id → orders.order_id`, `order_items.product_id → products.product_id`만 사용했습니다. |
| **개인정보**    | ✅ 통과 | 원본 데이터의 추가 행이나 개인정보 제공을 요구하지 않았습니다.                                                                                                                    |


# D. 실습 결과 기록 양식

## 66. 데이터 첫인상 보고서

# 데이터 첫인상 보고서

## 1. 데이터 파일

| 데이터셋        | 한 행의 의미           |   행 |  열 | 주요 ID         |
| ----------- | ----------------- | --: | -: | ------------- |
| customers   | 고객 1명의 정보         | 150 |  6 | customer_id   |
| products    | 상품 1개의 정보         | 100 |  4 | product_id    |
| orders      | 주문 1건의 정보         | 301 |  5 | order_id      |
| order_items | 주문에 포함된 상품 1개의 정보 | 764 |  5 | order_item_id |

## 2. 컬럼과 타입

* customers
  * customer_id(int64), name(str), gender(str), age(int64), city(str), signup_date(str)
* products
  * product_id(int64), product_name(str), category(str), price(int64)
* orders
  * order_id(int64), customer_id(int64), order_date(str), payment_method(str), order_status(str)
* order_items
  *order_item_id(int64), order_id(int64), product_id(int64), quantity(int64), unit_price(int64)

## 3. 결측치

* 모든 데이터셋에서 결측치 0건 확인

## 4. 전체 중복
* 모든 데이터셋에서 중복 행 0건 확인

## 5. 주요 ID 결측·중복
| ID            | 결측 | 중복 |
| ------------- | -: | -: |
| customer_id   | 0건 | 0건 |
| product_id    | 0건 | 0건 |
| order_id      | 0건 | 0건 |
| order_item_id | 0건 | 0건 |


## 6. 숫자형 범위
* 현재 단계에서는 숫자형 컬럼의 최소값·최대값 또는 이상치 범위는 확인하지 않았음.


## 7. 범주형 값
* 현재 단계에서는 범주형 컬럼의 고유값 및 일관성은 확인하지 않았음.


## 8. 날짜 범위와 변환 실패
* signup_date, order_date는 문자열(str) 타입임.
* 날짜 범위 및 변환 실패 여부는 현재 단계에서 확인하지 않았음.


## 9. 파일 간 참조 무결성
| 참조 관계                                          | 결과                            |
| ---------------------------------------------- | ----------------------------- |
| `orders.customer_id → customers.customer_id`   | 존재하지 않는 customer_id **1건 발견** |
| `order_items.order_id → orders.order_id`       | 이상 없음                         |
| `order_items.product_id → products.product_id` | 이상 없음                         |


## 10. 발견한 문제
* 현재 단계는 데이터 품질을 확인하는 단계이므로 데이터를 바로 삭제하거나 수정하지 않음.
* 외래키 무결성 오류의 원인을 확인한 후 다음 전처리 단계에서 처리 여부를 결정함.

## 11. 즉시 처리하지 않은 이유
* 현재 단계는 데이터 품질을 확인하는 단계이므로 데이터를 바로 삭제하거나 수정하지 않음.
* 외래키 무결성 오류의 원인을 확인한 후 다음 전처리 단계에서 처리 여부를 결정함.

## 12. 다음 전처리 단계에서 결정할 내용
* 외래키 무결성 오류(1건)의 처리 방법 결정
* signup_date, order_date의 날짜 타입 변환
* 숫자형 컬럼(age, price, quantity, unit_price)의 범위 및 이상치 확인
* 범주형 컬럼(gender, city, category, payment_method, order_status)의 값 일관성 확인

# F. 자습 과제

## 71. 과제 1. 네 파일 데이터 사전 작성

| 데이터셋            | 한 행의 의미           | 주요 ID           | 외래 키                                                                   | 숫자형                                                                 | 범주형                              | 날짜형            |
| --------------- | ----------------- | --------------- | ---------------------------------------------------------------------- | ------------------------------------------------------------------- | -------------------------------- | -------------- |
| **customers**   | 고객 1명의 정보         | `customer_id`   | 없음                                                                     | `customer_id`, `age`                                                | `name`, `gender`, `city`         | `signup_date`* |
| **products**    | 상품 1개의 정보         | `product_id`    | 없음                                                                     | `product_id`, `price`                                               | `product_name`, `category`       | 없음             |
| **orders**      | 주문 1건의 정보         | `order_id`      | `customer_id` → `customers.customer_id`                                | `order_id`, `customer_id`                                           | `payment_method`, `order_status` | `order_date`*  |
| **order_items** | 주문에 포함된 상품 1개의 정보 | `order_item_id` | `order_id` → `orders.order_id`<br>`product_id` → `products.product_id` | `order_item_id`, `order_id`, `product_id`, `quantity`, `unit_price` | 없음                               | 없음             |
